samples/oauth_rails_vethealth/app/models/authentication.rb

```diff
@@ -0,0 +1,53 @@
+class Authentication < ApplicationRecord
+  serialize :id_token
+  belongs_to :oauth_callback, optional: true
+
+  attr_reader :session_errors
+
+  def self.generate_nonce(base)
+    Digest::SHA256.hexdigest(base + ENV['va_developer_client_secret'])
+  end
+
+  def self.attributes_from_oauth(response)
+    attributes_array = response.map do |key,value|
+      clean_value =
+        case key
+        when 'expires_at'
+          Time.zone.at(value)
+        else
+          value
+        end
+      [key, clean_value]
+    end
+    Hash[attributes_array]
+  end
+
+  def parsed_id_token
+    @parsed_id_token ||= JWT.decode(self.id_token, nil, false).first
+  end
+
+  def expired?
+    @expired ||= self.expires_at < Time.zone.now
+  end
+
+  def authentic?(session)
+    @authentic ||= (self.parsed_id_token['nonce'] == Authentication.generate_nonce(session[:nonce_key]))
+  end
+
+  def validate_session(session)
+    @authentication_errors = []
+    if expired?
+      @authentication_errors << { type: :expired, message: 'The session has expired.' }
+    end
+    unless authentic?(session)
+      @authentication_errors << { type: :inauthentic, message: 'The token is not authentic!' }
+    end
+    @validated = true
+    @authentication_errors.empty?
```

Comment:
I'm not sure what you mean by this code. Does it mean that the validity of the session depends on whether or not there are errors? If so, what is the purpose of the `@validated` variable? Couldn't you just call `validate_session` from within `valid_session?` ?